### ISSUE TITLE
Extremely slow package resolution with min strategy

### ISSUE THREAD

**Issue by optical** [fsprojects/Paket#2442]
### Description

Using [this ](https://gist.github.com/optical/aa7610d21b039fab417b61ba03d315a7) paket.dependencies file, paket takes an extremely long time to perform a restore without a lock file. If you mix in `lowest_matching: true` into the mix it just about never completes (I've had it running for 20+ minutes now).

### Repro steps

Please provide the steps required to reproduce the problem

1. Download this dependencies file: https://gist.github.com/optical/aa7610d21b039fab417b61ba03d315a7
2. Run paket install
3. Note that it takes 3+ minutes to resolve (2m30s on runtime itself)

Test two:

1. Download this dependencies file: https://gist.github.com/optical/aa7610d21b039fab417b61ba03d315a7
2. Add `lowest_matching: true` to the file
3. Run paket install
4. Go home for the day and see if its finished tomorrow

### Expected behavior
Resolving with `strategy: min` and `lowest_matching: true` should succeed within a reasonable timeframe

**Comment by forki** [comment]
One way to get rid of the problem is usually to remove some of the package
version restrictions. Your are pinning a lot of the versions so many many
resolutions are forbidden. Usually things resolve much faster when you
start to remove some of the pins.

Am 19.06.2017 12:56 vorm. schrieb "Jared Klopper" <notifications@github.com

**Comment by optical** [comment]
What do you mean by pinning? All of the top-level constraints use the twiddle wakka symbol, which should specify a reasonable semver constraint. on either minor or patch level. Is there a better approach we should be taking when specifying constraints?

**Comment by matthid** [comment]
Is this a regression. I mean did this work in paket 4?

**Comment by forki** [comment]
@optical yes that is what it does but I assume using that Newtonsoft.json range is a problem since many of the other packages are probably incompatible. So it's just having a hard time to find something that works

**Comment by matthid** [comment]
What I'm asking myself. 
If you use `lowest_matching` and `min` strategy what are you actually doing?
You basically do updates manually and use paket only to find a valid resolution?

But yes we should look into this, IMHO it should work :/

**Comment by forki** [comment]
@matthid that is full nuget compat mode if you will

**Comment by matthid** [comment]
@forki not sure about that as you have no "update" ...

**Comment by optical** [comment]
@matthid more or less, yes. Our usage is complex and non-typical. We built our own in-house package/module management system many years ago - long before Nuget etc. What we now have is many shared source components, that them selves can depend upon Nuget packages. We then have many different applications using these shared components. You can imagine each component as having a list of components it depends on, as well as a set of nuget packages using pakets version constraints.

Our usage of paket within this system is entirely automated. All component source code is included into a single generated csproj during build. A paket.dependencies file is then generated based off the components, restore is done, the dependencies file is erased and the build continues. The generated `paket.lock` file is then kept around to be checked into source code. This process means we can do resolve quite often as many components change frequently, so many applications paket.lock will change quite often (versioning of our components is bad, very little control over upgrades).

We have quite a lot of applications and even more components, and we ran into trouble using the default paket resolve strategies initially.

@matthid I'm testing with 4.8.8 at the moment. It does appear to be super slow too, but it doesn't spam the same line repeatedly, so its much easier to see why paket is getting a little stuck, which is more helpful as I could adjust the constraints and it'll arrive at the solution quicker. But it doesn't appear to be a perf regression.

**Comment by matthid** [comment]
>  All component source code is included into a single generated csproj 

So the versions are already given? Why don't you create the lockfile directly? Or am I missing something here.

**Comment by optical** [comment]
Its an interesting Idea I haven't considered. That would require walking the full dependency graph - we just specify top level constraints against packages the component is interested in, and let paket take a full look at the picture by unpicking the mess that is all the transitive dependencies, the majority of which are not specified, and in this particular case appear to be the problem.

**Comment by matthid** [comment]
So, but if my assumption is correct you should use `=` or `==`, right? because paket can only really tell you if there are problems. It should not change any versions, correct?

**Comment by optical** [comment]
Not quite, we can float in two main scenarios. The primary situation is
when we depend on something like Newtonsoft ~> 6.0, but we also depend on
SomeOtherPackage, which in turn asks for Newtonsoft at say 6.0.5. In this
case we get 6.0.5 despite all our known dependencies saying 6.0.0.

The other situation is when two components request differing version
constraints for the same package. It means our paket dependencies file has
duplicate declarations​ for the same package, but differing constraints.
This isn't a big deal and something that I wouldn't worry about since we
can easily address this internally.

**Comment by forki** [comment]
Also the problems come because of unspecified the transitives.

**Comment by matthid** [comment]
This is probably just a matter or sorting some lists differently in the resolver on this particular configuration.
I can see that it runs into newer versions than it needs to and then "never" finds a solution.

**Comment by careylu11** [comment]
I'm not sure if this would be the right thread to post this, but I'm having what appears to be similar issues. A paket install on my solution (converted from nuget using the convert-from-nuget command) took over 7 hours to tell me that there was a conflict that needed resolution. Is there anything I can do to speed this up?

**Comment by forki** [comment]
no @careylu11 if resolution windows are too small then paket is usually very slow and you need to open the restrictions a bit. @isaacabraham wrote a guide: 

* https://cockneycoder.wordpress.com/2017/08/14/migration-basics-from-nuget-to-paket/ 
* https://cockneycoder.wordpress.com/2017/08/21/migrating-complex-nuget-solutions-to-paket/

**Comment by careylu11** [comment]
Awesome! Exactly what I was looking for. Thanks for the quick response @forki.
